### PR TITLE
`CMS_decrypt*()`: fix misconceptions and memory leak - backport to v3.0 and 3.1

### PR DESCRIPTION
**Pull request by DDvO**
This is a backport of #19222 as motivated by https://github.com/openssl/openssl/pull/19222#issuecomment-1248204946.
It also includes the fixup done in #19981.

### CODE REVIEW

**Review by t8m — APPROVED** [review]

File: doc/man3/CMS_decrypt.pod [hunk 1]
```diff
@@ -14,31 +15,48 @@ content from a CMS envelopedData structure
...
+provides the content unless I<cert>, I<dcont>, and I<out> are NULL as well.
+The I<dcont> parameter is used in the rare case where the encrypted content
 is detached. It will normally be set to NULL.
 
-CMS_decrypt_set1_pkey_and_peer() associates the private key B<pkey>, the
-corresponding certificate B<cert> and the originator certificate B<peer> with
-the CMS_ContentInfo structure B<cms>.
+CMS_decrypt_set1_pkey_and_peer() decrypts the CMS_ContentInfo structure I<cms>
+using the private key I<pkey>, the corresponding certificate I<cert>, which is
+recommended but may be NULL, and the (optional) originator certificate I<peer>.
```
Comment (paulidale) [inline doc/man3/CMS_decrypt.pod:39]: which is recommended to be supplied but can be NULL

File: doc/man3/CMS_decrypt.pod [hunk 2]
```diff
@@ -14,31 +15,48 @@ content from a CMS envelopedData structure
...
+On success, it also records in I<cms> the decryption key I<pkey>, and then
```
Comment (paulidale) [inline doc/man3/CMS_decrypt.pod:40]: s/then/this/

File: doc/man3/CMS_decrypt.pod [hunk 3]
```diff
@@ -14,31 +15,48 @@ content from a CMS envelopedData structure
...
+This call deallocates any decryption key stored in I<cms>.
+
+CMS_decrypt_set1_pkey() is the same as
+CMS_decrypt_set1_pkey_and_peer() with I<peer> being NULL.
 
-CMS_decrypt_set1_pkey() associates the private key B<pkey>, corresponding
-certificate B<cert> with the CMS_ContentInfo structure B<cms>.
+CMS_decrypt_set1_password() decrypts the CMS_ContentInfo structure I<cms>
+using the secret I<pass> of length I<passlen>.
+On success, it also records in I<cms> the decryption key used, and then
```
Comment (paulidale) [inline doc/man3/CMS_decrypt.pod:49]: s/then/this/

File: doc/man3/CMS_decrypt.pod [hunk 4]
```diff
@@ -54,26 +72,32 @@ open to attack.
...
+or use CMS_decrypt_set1_password() if the recipient has a symmetric key.
+In these cases both I<cert> and I<pkey> should be set to NULL.
 
 To process KEKRecipientInfo types CMS_set1_key() or CMS_RecipientInfo_set0_key()
 and CMS_RecipientInfo_decrypt() should be called before CMS_decrypt() and
-B<cert> and B<pkey> set to NULL.
+I<cert> and I<pkey> set to NULL.
 
-The following flags can be passed in the B<flags> parameter.
+The following flags can be passed in the I<flags> parameter.
```
Comment (paulidale) [inline doc/man3/CMS_decrypt.pod:83]: Is it worth adding a list of the flags here?  The descriptions could either be inline or later.

It seems like a list should go here but there isn't one.

File: doc/man3/CMS_decrypt.pod [hunk 5]
```diff
@@ -14,31 +15,48 @@ content from a CMS envelopedData structure
...
-or AuthEnvelopedData structure. B<pkey> is the private key of the recipient,
-B<cert> is the recipient's certificate, B<out> is a BIO to write the content to
-and B<flags> is an optional set of flags.
-
-The B<dcont> parameter is used in the rare case where the encrypted content
+CMS_decrypt() extracts the decrypted content from a CMS EnvelopedData
+or AuthEnvelopedData structure.
+It uses CMS_decrypt_set1_pkey() to decrypt the content
+with the recipient private key I<pkey> if I<pkey> is not NULL.
+In this case, the associated certificate is recommended to provide in I<cert> -
```
Comment (paulidale) [inline doc/man3/CMS_decrypt.pod:27]: It's not clear what is meant here.  Specifically the _recommended to provide in I<cert>_ part.

File: doc/man3/CMS_decrypt.pod (hunk 2 above)
Comment (DDvO) [reply doc/man3/CMS_decrypt.pod]: Changed, here as well as in the other occurrence.

File: doc/man3/CMS_decrypt.pod (hunk 5 above)
Comment (DDvO) [reply doc/man3/CMS_decrypt.pod]: The grammar of the sentence was a bit weird - improved.

File: doc/man3/CMS_decrypt.pod (hunk 4 above)
Comment (DDvO) [reply doc/man3/CMS_decrypt.pod:84]: > Is it worth adding a list of the flags here? The descriptions could either be inline or later.
> 
> It seems like a list should go here but there isn't one.

I agree. Yet I do not know whether further flags apply in this case.
Here my aim was just to update the symbol layout.
The original text was given by @richsalz  in 99d63d4662e16afbeff49f29b48f1c87d5558ed0.

**Comment by DDvO** [comment]
> AFAICT, the CI failures are not relevant?

Yep, since months I see one or two if those Windows runs fail on most PRs.

**Comment by DDvO** [comment]
Merged to 3.0 and cherry-picked to 3.1 - thanks @t8m and @paulidale for the approvals.

**Pull request closed without merging**